### ISSUE TITLE
How to evaluate MoveNet.SinglePose.Lightning using COCOEval API?

### ISSUE THREAD

**Issue by ozwis** [tensorflow/hub#876]
### What happened?

I’m trying to evaluate [MoveNet.SinglePose.Lightning](https://tfhub.dev/google/movenet/singlepose/lightning/4) using [the COCOEval API](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py).  
However, I do not know how to calculate the score to use for evaluation.

Referring to [the Results Format page](https://cocodataset.org/#format-results), one score must be set for each inference result for the evaluation of keypoint detecition.
And this score is the instance-level confidence of the object.

```python

[{
  "image_id": int,
  "category_id": int,
  "keypoints": [x1,y1,v1,...,xk,yk,vk],
  "score": float,
}]

```

But, MoveNet.SinglePose.Lightning outputs the confidence for each keypoint and not the instance-level confidence.

ref. [The model card](https://storage.googleapis.com/movenet/MoveNet.SinglePose%20Model%20Card.pdf)

```shell
Outputs

A float32 tensor of shape [1, 1, 17, 3].

    The first two channels of the last dimension represents the yx coordinates (normalized to image frame, i.e. range in [0.0, 1.0]) of the 17 keypoints (in the order of: [nose, left eye, right eye, left ear, right ear, left shoulder, right shoulder, left elbow, right elbow, left wrist, right wrist, left hip, right hip, left knee, right knee, left ankle, right ankle]).

    The third channel of the last dimension represents the prediction confidence scores of each keypoint, also in the range [0.0, 1.0].
```

How can I calculate the score or do the evaluation?

### tensorflow_hub Version

0.12.0 (latest stable release)

### TensorFlow Version

2.8 (latest stable release)

### Python Version

3.8

### OS

Linux

**Comment by singhniraj08** [comment]
@ozwis,

`movenet/singlepose/lightning` outputs the confidence of each keypoints and doesn't outputs the confidence score of the instance. 

For confidence score of the instance, you can use [movenet/multipose/lightning](https://tfhub.dev/google/movenet/multipose/lightning/1) which generates the keypoint location and confidence scores, the region of the bounding box and the confidence score of the instance.

Thank you!

> 
> **Outputs generated from multipose lightning:**
> A float32 tensor of shape [1, 6, 56].
> 
> - The first dimension is the batch dimension, which is always equal to 1.
> - The second dimension corresponds to the maximum number of instance detections. The model can detect up to 6 people in the image frame simultaneously.
> - The third dimension represents the predicted bounding box/keypoint locations and scores. The first 17 * 3 elements are the keypoint locations and scores in the format: [y_0, x_0, s_0, y_1, x_1, s_1, …, y_16, x_16, s_16], where y_i, x_i, s_i are the yx-coordinates (normalized to image frame, e.g. range in [0.0, 1.0]) and confidence scores of the i-th joint correspondingly. The order of the 17 keypoint joints is: [nose, left eye, right eye, left ear, right ear, left shoulder, right shoulder, left elbow, right elbow, left wrist, right wrist, left hip, right hip, left knee, right knee, left ankle, right ankle]. The remaining 5 elements [ymin, xmin, ymax, xmax, score] represent the region of the bounding box (in normalized coordinates) and the confidence score of the instance.

**Comment by ozwis** [comment]
@singhniraj08 
Thank you for quick response!

> For confidence score of the instance, you can use [movenet/multipose/lightning](https://tfhub.dev/google/movenet/multipose/lightning/1) which generates the keypoint location and confidence scores, the region of the bounding box and the confidence score of the instance.

So, in order to evaluate the `singlepose` model with the COCOEval API, I should use the output of `movenet/singlepose/lighting` as keypoints and the output of `movenet/multipose/lighting` as the instance confidence score.

I will try it out. Thanks!

**Comment by singhniraj08** [comment]
@ozwis,

Could you please close this issue, if the issue has been resolved. Thank you!

**Comment by ozwis** [comment]
@singhniraj08   
Sorry for late reply.

I tested `movenet/singlepose/lightning` , got a result **mAP=0.399**.

  ```bash
  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.399
  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.680
  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.404
  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.088
  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.494
  Average Recall     (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.451
  Average Recall     (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.716
  Average Recall     (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.465
  Average Recall     (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.128
  Average Recall     (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.554
  ```

To reproduce the evaluation dataset described in the [model card,](https://storage.googleapis.com/movenet/MoveNet.SinglePose%20Model%20Card.pdf)   
I created a subset of the COCO Keypoint Dataset Validation Set 2017, according to the following rules.
* Extracted data that have only one annotation in person_keypoints_val2017.json.
  * 1045 out of 5k data were extracted.
* Excluded 21 face-only, 80 hand-only, 21 foot-only, and 4 data with unclear annotations.
  * 919 out of 1045 data remained.
  * The list of 919 filenames: [val2017_sigle_person_919.txt](https://github.com/tensorflow/hub/files/10785521/val2017_sigle_person_919.txt)
* Excluded 2 gray scale data.
  * 917 data remained.  
So, I used this 917 data for my evaluation dataset.

Although the model cards show evaluation results for different attributes/categories and thus cannot be strictly compared, the expected results (mAP of around 60+%) were not achieved.

Did I make a mistake?

**Comment by singhniraj08** [comment]
@ozwis,

Apologies for late reply. Please refer to this [article](https://medium.com/@timothycarlen/understanding-the-map-evaluation-metric-for-object-detection-a07fe6962cf3) and [code](https://gist.github.com/tarlen5/008809c3decf19313de216b9208f3734) to calculate `mAP` for `movenet/singlepose/lighting`. 
Hope this helps. Thank you!

**Comment by yozawa-sony** [comment]
@singhniraj08 
Thanks for the guidance.
I checked the articles and they seem to be about object detection using bounding box, not keypoint detection as `movenet/singlepose/lighting` does.  
(Of course, I see no difference between these two task evaluations, except for the use of OKS instead of IoU as the evaluation metric)

**Comment by ozwis** [comment]
@singhniraj08   
Thanks for the guidance.  
I checked the articles and they seem to be about object detection using bounding box , not keypoint detection as `movenet/singlepose/lighting` does.  
(Of cause, I see no difference between these two task evaluations, except for the use of OKS instead of IoU as the evaluation metric)  

Anyway, do you know if the evaluation code or more detail of the evaluation dataset is available anywhere?

**Comment by singhniraj08** [comment]
@ozwis,

Apologies, I couldn't find evaluation code for key point detection. Mostly mAP is used for object detection tasks which in this case would work with instance level confidence score.

As your question is not a bug/performance/feature request, I would recommend you to open this issue in [Stackoverflow Forum](https://stackoverflow.com/) as there is a larger community there. Thanks!

**Comment by ozwis** [comment]
@singhniraj08  
Thanks for all your help!  
Actually, I've already posted [a question](https://stackoverflow.com/questions/75386590/how-to-evaluate-movenet-singlepose-lightning-using-cocoeval-api), but it doesn't look like I'll get a response.  
Since there seems to be no further progress, I'll close this issue.
